File: MultiThread/wordcount.py
import os
import multiprocessing as mp
import mmap

def count_words_in_chunk(filename, start, end):
    with open(filename, 'r', encoding='utf-8') as f:
        # Memory-map the file
        mm = mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ)
        # Move to the start offset
        mm.seek(start)
        if start != 0:
            # Move to the next whitespace to avoid splitting a word
            while mm.tell() < end:
                if chr(mm.read_byte()).isspace():
                    break
        actual_start = mm.tell()
        mm.seek(actual_start)
        size = min(end, mm.size())
        text = mm.read(size - actual_start)
        # Only read up to `end`
        if mm.tell() > end:
            text = text[:end - actual_start]
        # Count words in this chunk
        word_count = len(text.decode('utf-8', errors='ignore').split())
        mm.close()
        return word_count

def parallel_word_count(filename, n_processes=None):
    if n_processes is None:
        n_processes = os.cpu_count() or 2
    filesize = os.path.getsize(filename)
    # Divide file into chunks
    chunk_size = filesize // n_processes
    offsets = []
    for i in range(n_processes):
        start = i * chunk_size
        end = filesize if i == n_processes - 1 else (i + 1) * chunk_size
        offsets.append((start, end))

    with mp.Pool(n_processes) as pool:
        results = pool.starmap(count_words_in_chunk, [(filename, start, end) for (start, end) in offsets])
    return sum(results)

if __name__ == "__main__":
    import time
    filename = "../largefile.txt"
    start = time.time()
    total = parallel_word_count(filename)
    end = time.time()
    print("Word count:", total)
    print("Execution time: %.2f seconds" % (end - start))

File: MultiThread/wordcount.rs
use std::fs::File;
use std::io::{Read, Seek, SeekFrom};
use std::thread;

fn count_words_in_chunk(filename: &str, start: u64, end: u64) -> usize {
    let mut file = File::open(filename).expect("open failed");
    file.seek(SeekFrom::Start(start)).expect("seek failed");
    let mut buf = vec![0u8; (end - start) as usize];
    file.read_exact(&mut buf).expect("read failed");
    let text = String::from_utf8_lossy(&buf);

    // For all except the first chunk, skip to the first whitespace (to avoid partial word)
    let slice = if start != 0 {
        match text.find(|c: char| c.is_whitespace()) {
            Some(pos) => &text[pos..],
            None => "",
        }
    } else {
        &text
    };

    slice.split_whitespace().count()
}

fn main() {
    let filename = "../largefile.txt";
    let metadata = std::fs::metadata(filename).expect("metadata failed");
    let file_size = metadata.len();
    let n_threads = std::thread::available_parallelism().map_or(1, |n| n.get());
    let mut handles = Vec::new();

    let chunk_size = (file_size as usize + n_threads - 1) / n_threads;

    let mut start: u64 = 0;
    for _ in 0..n_threads {
        let end = std::cmp::min(start + chunk_size as u64, file_size);
        let fname = filename.to_string();
        handles.push(thread::spawn(move || count_words_in_chunk(&fname, start, end)));
        start = end;
        if start >= file_size {
            break;
        }
    }

    let now = std::time::Instant::now();
    let total: usize = handles.into_iter().map(|h| h.join().unwrap()).sum();
    let elapsed = now.elapsed();

    println!("Word count: {}", total);
    println!("Execution time: {:.2} seconds", elapsed.as_secs_f64());
}
